pkg/session/config.go

```diff
@@ -23,40 +26,119 @@ type Config struct {
 	// it takes to connect to a storage provider, it is used to determine the
 	// prioritisation of storage providers. The value determines the weight of
 	// previous connect times, a lower value will give more weight to recent
 	// connect times. A value of 0 will only use the most recent connect time.
 	ConnectTimeAlpha float64
+	// OverallConnectTimeAlpha is the alpha value for the exponential moving
+	// average of the overall connect time. The overall connect time is the
+	// average connect time of *all* storage providers, used to normalise the
+	// connect time metric.
+	OverallConnectTimeAlpha float64
+	// SuccessAlpha is the alpha value for the exponential moving average of
+	// the success metric for a storage provider. The success metric is the
+	// ratio of successful retrievals to total retrievals. The value determines
+	// the weight of previous success metrics, a lower value will give more
+	// weight to recent success metrics. A value of 0 will only use the most
+	// recent success metric.
+	SuccessAlpha float64
+
+	// Random is an optional rng, if nil, math/rand will be used.
+	Random Random
+
+	VerifiedDealWeight  float64
+	FastRetrievalWeight float64
+	ConnectTimeWeight   float64
+	SuccessWeight       float64
 }
 
+// DefaultConfig returns a default config with usable alpha and weight values.
 func DefaultConfig() *Config {
 	return &Config{
-		MaxFailuresBeforeSuspend: 5,
-		SuspensionDuration:       time.Minute * 10,
-		FailureHistoryDuration:   time.Second * 30,
-		ConnectTimeAlpha:         0.5, // only use most recent connect time
+		ConnectTimeAlpha:        0.5,
+		OverallConnectTimeAlpha: 0.8,
+		SuccessAlpha:            0.5,
+		VerifiedDealWeight:      3.0,
+		FastRetrievalWeight:     2.0,
+		ConnectTimeWeight:       1.0,
+		SuccessWeight:           1.0,
+	}
+}
+
+// WithProviderBlockList sets the provider blocklist.
+func (cfg Config) WithProviderBlockList(blocklist map[peer.ID]bool) *Config {
+	cfg.ProviderBlockList = blocklist
+	return &cfg
+}
+
+// WithProviderAllowList sets the provider allowlist.
+func (cfg Config) WithProviderAllowList(allowlist map[peer.ID]bool) *Config {
+	cfg.ProviderAllowList = allowlist
+	return &cfg
+}
+
+// WithDefaultProviderConfig sets the default provider config.
+func (cfg Config) WithDefaultProviderConfig(providerConfig ProviderConfig) *Config {
+	cfg.DefaultProviderConfig = providerConfig
+	return &cfg
+}
+
+// WithConnectTimeAlpha sets the connect time alpha.
+func (cfg Config) WithConnectTimeAlpha(alpha float64) *Config {
+	cfg.ConnectTimeAlpha = alpha
+	return &cfg
+}
+
+// WithOverallConnectTimeAlpha sets the overall connect time alpha.
+func (cfg Config) WithOverallConnectTimeAlpha(alpha float64) *Config {
+	cfg.OverallConnectTimeAlpha = alpha
+	return &cfg
+}
+
+// WithSuccessAlpha sets the success alpha.
+func (cfg Config) WithSuccessAlpha(alpha float64) *Config {
+	cfg.SuccessAlpha = alpha
+	return &cfg
+}
+
```

Comment:
Probably want configs to adjust the weights no?

Reply:
or I mean helper functions, if this is what we're generally going with

Reply:
I'm open to suggestions here btw, I added these cause I realised that you almost never want a `&Config{...}`, you want a `DefaultConfig()` and then customise that, because other wise weights and parameters are all wrong. So these methods are all just sugar.
It's a bit of a foot-gun allowing a bare instantiation but I don't really want to make the variables private.